packages/itwinui-react/src/react-table/react-table-types.test.tsx

```diff
@@ -0,0 +1,802 @@
+                ids: ['111'],
+                startDate: new Date('May 11, 2021'),
+                endDate: '2021-05-31T21:00:00.000Z',
+                subRows: [],
+              },
+            ],
+          },
+        ],
+      },
+      {
+        index: 2,
+        name: 'Name2',
+        description: 'Description2',
+        ids: ['2', '3', '4'],
+        startDate: new Date('May 2, 2021'),
+        endDate: '2021-06-01T21:00:00.000Z',
+        subRows: [
+          {
+            index: 21,
+            name: 'Name21',
+            description: 'Description21',
+            ids: ['21'],
+            startDate: new Date('May 21, 2021'),
+            endDate: '2021-06-01T21:00:00.000Z',
+            subRows: [],
+          },
+          {
+            index: 22,
+            name: 'Name22',
+            description: 'Description22',
+            ids: ['22'],
+            startDate: new Date('May 22, 2021'),
+            endDate: '2021-06-01T21:00:00.000Z',
+            subRows: [],
+          },
+        ],
+      },
+      {
+        index: 3,
+        name: 'Name3',
+        description: 'Description3',
+        ids: ['3', '4'],
+        startDate: new Date('May 3, 2021'),
+        endDate: '2021-06-02T21:00:00.000Z',
+        subRows: [],
+      },
+    ],
+    [],
+  ) satisfies TableRowDataType[];
+
+  const onFilter = React.useCallback(
+    (
+      filters: TableFilterValue<TableRowDataType>[],
+      state: TableTypes.TableState,
+      filteredData: TableTypes.Row<TableRowDataType>[],
+    ) => {
+      // rowInfo is used due to JSON errors when displaying row data
+      let rowInfo = '[';
+      filteredData.forEach((row) => {
+        rowInfo += `${JSON.stringify(row.original)},`;
+      });
+      rowInfo = rowInfo.slice(0, rowInfo.length - 1);
+      rowInfo += ']';
+
+      console.log(
+        `Filter changed. Filters: ${JSON.stringify(
+          filters,
+        )}, State: ${JSON.stringify(state)}, Rows: ${rowInfo}`,
+      );
+    },
+    [],
+  ) satisfies NonNullable<TableProps<TableRowDataType>['onFilter']>;
+
+  const onSelect = React.useCallback(
+    (rows, state) =>
+      console.log(
+        `Selected rows: ${JSON.stringify(rows)}, Table state: ${JSON.stringify(
+          state,
+        )}`,
+      ),
+    [],
+  ) satisfies NonNullable<
+    TableProps<{ name: string; description: string }>['onSelect']
+  >;
+
+  const onRowClick = React.useCallback(
+    (event: React.MouseEvent, row: TableTypes.Row) =>
+      console.log(`Row clicked: ${JSON.stringify(row.original)}`),
+    [],
+  );
+
+  const onClickHandler = (
+    props: TableTypes.CellProps<{ name: string; description: string }>,
+  ) => console.log(props.row.original.name);
+
+  const onSort = React.useCallback(
+    (state) =>
+      console.log(`Sort changed. Table state: ${JSON.stringify(state)}`),
+    [],
+  ) satisfies NonNullable<
+    TableProps<{ name: string; description: string }>['onSort']
```

Comment:
i noticed many instances of `{ name: string; description: string }` in this particular test. shouldn't test be `TableRowDataType`?

Reply:
I noticed that before `Table.stories.tsx` had fixed the type errors, there were many type assertions to a subset of the fields. E.g. If `TableRowDataType` has `name`, `description`, and a few more fields, if `onSort` would only care about name and description, the type assertions were with this subset object instead of with `TableRowDataType`.

So I thought that it would be good to confirm that the types work with this subset too. i.e. when we pass this `onSort` function (that is type asserted to using the subset object) to `<Table<TableRowDataType> ...>`, we don't get any error.

I noticed that I didn't pass any generic to `<Table>`. In https://github.com/iTwin/iTwinUI/pull/1514/commits/21d1c0f376028ea04eb850d2251e03749ff14dfb, I did `<Table<TableRowDataType> ...>` to test the types more strictly.

Reply:
> If `TableRowDataType` has `name`, `description`, and a few more fields, if `onSort` would only care about name and description, the type assertions were with this subset object instead of with `TableRowDataType`

but now the fields are duplicated and error-prone. i don't see a downside to using `TableRowDataType` everywhere; any fields that are not pertinent are not getting in the way

---

> I noticed that I didn't pass any generic to `<Table>`

this is unrelated and unnecessary. it should be automatically inferred

Reply:
> > I noticed that I didn't pass any generic to `<Table>`
> 
> this is unrelated and unnecessary. it should be automatically inferred

I meant that if we don't pass the generic type to `<Table>`, it the generic becomes Record<string, unknown>

<!-- ![image](https://github.com/iTwin/iTwinUI/assets/45748283/537c60b2-677f-4b77-9b46-3da06c81f39b) -->
![image](https://github.com/iTwin/iTwinUI/assets/45748283/cc1ae249-d65d-470c-89b5-0dfd9c22f665)

But if we pass the generic type, then the props change to use the stricter data type.

![image](https://github.com/iTwin/iTwinUI/assets/45748283/5e8386ee-69ca-4636-9765-bc18daa11559)

So I thought it would be worth confirming if type errors are thrown when the Table expects a strict type and we pass an `onSort` that uses a subset of that strict type (i.e. only `name` and `description`).

---

But looks like even if the onSort uses an incorrect generic type (`{ foo: number, bar: number} ` instead of `{ name:string, description: string }`), no error is thrown in `<Table<TableRowDataType> ...>`.

![image](https://github.com/iTwin/iTwinUI/assets/45748283/e5eaafd0-6d17-427d-b2df-81ab7630a3f6)

https://github.com/iTwin/iTwinUI/pull/1514/commits/d0c2a6a1dc49c0db4ecf7e8c03a83a84886c9b66: So I changed all subsets to use `TableRowDataType` instead. I also intentionally didn't change some `(typeof data)[number]` to `TableRowDataType`. I felt it would be more robust to test multiple approaches of passing types.

Reply:
> I meant that if we don't pass the generic type to `<Table>`, it the generic becomes Record<string, unknown>

this seems like a bug tbh. the generic type of Table should be inferred from what is passed into columns and/or data

Reply:
Sounds good, for now I removed that generic type passed to `<Table>`. (https://github.com/iTwin/iTwinUI/pull/1514/commits/0007c277056a8965bd2d37bd0eb64e441b2288be)

```diff
- <Table<TableRowDataType>
+ <Table
  ...
/>
```

Reply:
don't these still need to be changed?

<img width="1096" alt="" src="https://github.com/iTwin/iTwinUI/assets/9084735/c31c77fb-9eb7-4f7b-8689-6302ec0cdd9b">

Reply:
Changed (https://github.com/iTwin/iTwinUI/pull/1514/commits/fdff0a98f9c4198b434e37595b1deabadfba2133)